deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -52,25 +55,178 @@ def upload(options, screenshots, max_n_threads = 16)
 
-      upload_screenshots(screenshots_per_language, localizations, options, max_n_threads)
+      upload_screenshots(updated_sets_per_language, localizations, options, max_n_threads)
     end
 
-    def delete_screenshots(localizations, screenshots_per_language, max_n_threads, tries: 5)
+    def get_updated_screenshots(localizations, screenshots_per_language)
+      sets_per_language = {} # per locale and device type; all provided screenshots
+      updated_sets_per_language = {} # per locale and device type; just the updated screenshots
+
+      # first, divide the new screenshots into sets
+      screenshots_per_language.each do |language, screenshots_for_language|
+        screenshots_per_device_type = {}
+
+        screenshots_for_language.each do |screenshot|
+          display_type = screenshot.device_type
+
+          if display_type.nil?
+            UI.error("Error... Screenshot size #{screenshot.screen_size} not valid for App Store Connect")
+            next
+          end
+
+          bytes = File.binread(screenshot.path)
+          checksum = Digest::MD5.hexdigest(bytes)
+
+          screenshots_per_device_type[display_type] ||= []
+          screenshots_per_device_type[display_type] << {
+              screenshot: screenshot,
+              checksum: checksum
+          }
+        end
+
+        sets_per_language[language] = screenshots_per_device_type
+      end
+
+      # then, compare the new screenshots with the existing ones
+      sets_per_language.each do |language, screenshots_per_device_type|
+        updated_screenshots_per_device_type = {}
+        localization = localizations.find do |l|
+          l.locale == language
+        end
+
+        unless localization
+          UI.error("Couldn't find localization on version for #{language}")
+          next
+        end
+
+        app_screenshot_sets_map = {}
+        app_screenshot_sets = localization.get_app_screenshot_sets
+        app_screenshot_sets.each do |app_screenshot_set|
+          app_screenshot_sets_map[app_screenshot_set.screenshot_display_type] = app_screenshot_set
+        end
+
+        screenshots_per_device_type.each do |device_type, screenshots_with_checksums|
+          existing_screenshots = app_screenshot_sets_map[device_type].app_screenshots
+          updated_screenshots_per_device_type[device_type] ||= []
+
+          temp_checksums = []
+          screenshots_with_checksums.each_with_index do |screenshot_with_checksum, index|
+            if index >= 10
+              UI.error("Too many screenshots found for device '#{device_type}' in '#{language}', skipping this one (#{screenshot_with_checksum[:screenshot].path})")
+              next
+            end
+
+            # skip over duplicates
+            if temp_checksums.include?(screenshot_with_checksum[:checksum])
+              UI.message("Duplicate found. Skipping '#{screenshot.path}'...")
+              next
+            end
+            temp_checksums << screenshot_with_checksum[:checksum]
```

Comment:
Corner case: what happens to the update mechanism in case a screenshot is duplicated? Since we're matching index-by-index, we might end up skipping one check 🤔 

Consider the following example:

- App Store Connect set: [1, 2, 3, 4, 5]
- Candidate new set: [1, 2, 2, 4, 5]

The end result would be that:

- 1 and 2 are preserved since they match the corresponding positions on App Store Connect
- the second iteration of 2 gets skipped
- 4 and 5 are preserved since they match the corresponding positions on App Store Connect

Although `3` is not present in the source set, it will still be present on App Store Connect, because it won't be compared with the corresponding local screenshot (since that screenshot has been skipped). 

I'm not 100% sure of what the "right" thing to do would be in this case. Considering that there's no good reason for ASO to upload the same screenshot twice _and_ that they don't want for sure to have zombie 🧟 screenshots on App Store Connect, maybe we could run a local duplication check across all files before even attempting the upload, and stop the upload if a duplicate is detected. 

Another option could be to upload the duplicate screenshot regardless (in order to replace the 🧟 screenshot), verify the checksums after the upload, and delete duplicates from App Store Connect _a posteriori_. This _should_ be safe to do, and perhaps it doesn't even add extra network calls since we know the checksums _a priori_ (we're just postponing the moment when we'll actually check for duplicates, and adding an extra _delete_ for each duplicate in the worst case).

Reply:
Yes, the logic you described is correct. 😃 If we want to be more like the original screenshot uploading logic that you described, we could do something like this:
- Detect that the `3` is not present anymore -> mark it for removal
- Don't upload the second `2` since it is a duplicate
The end result would be: 1, 2, 4, 5.
I have to check how this could be done if we like it. Otherwise, any of the two solutions work too. I'd say that stopping everything because of a duplicate is a bit drastic however it could be a good idea since, like you said, we are 99% sure ASO would never want a duplicate screenshot to be present (we could ask them to make sure).

Reply:
I'm also fine with your proposal, as long as we find a way to detect that 3 has gone missing and that should be deleted. We'd need to be a bit careful with the indexes though, because what you'd want to do at that point is either: 

1. complete the upload (without the second `2`) and _then_ delete `3`, otherwise you mess up the indexes (there'd be a gap in the middle), or
2. delete `3` immediately after the comparison and decrement all subsequent positions before the upload.

Reply:
Okay, I have made some changes here:
1. We will first scan for duplicates and remove every duplicate for the same language and device type. This is now done before determining the positions, so there are no issues with the index increase during the looping.
2. We delete screenshots that no longer need to exist (so, not only the screenshots that need to be replaced are deleted).

With this, what would happen in this corner case would be the following:
- In App Store, the screenshots are [1, 2, 3, 4, 5].
- [1, 2, 2, 4, 5] are supplied for the metadata update.
- Duplicates are removed, resulting with the expected screenshots being [1, 2, 4, 5].
- 1 and 2 remain untouched.
- 3 is deleted and will be replaced by a 4.
- 4 is deleted and will be replaced by a 5
- 5 is just deleted.
- The end state is the expected [1, 2, 4, 5].